### PR TITLE
Use layerwise number of channels and size for normalizing style loss

### PR DESCRIPTION
**Pull request by sjosund**
Following https://arxiv.org/pdf/1508.06576.pdf we extract the normalizing constants for each layer.

### CODE REVIEW

**Comment by Dref360** [comment]
Maybe add a comment on why you're doing this. Does this change the result a lot? If so, update https://github.com/sjosund/keras/blob/46c9d13563aba642c2bfe8e4158e746e1acc51ac/examples/neural_style_transfer.py#L21.

**Comment by fchollet** [comment]
Can you post a before / after image with the same settings? E.g. starry night + tuebingen.

**Comment by titu1994** [comment]
I tried out the changes that were suggested in the commit, and couldn't detect much of a change in the output.

The one on the top is the current script (without the change), the one on the bottom is with the change :
current_setting_at_iteration_75 new_setting_at_iteration_75

**Comment by fchollet** [comment]
Which reference images did you use? Last time I tried the script with some image I got off of a Google search, it didn't look as good.

**Comment by titu1994** [comment]
Tubingen and Starry night were from the net - 
tubingen

The Starry Night - 
starry_night

I think the difference may have been due to my use of VGG 19 instead of VGG 16, and the total variation regularization weight of 8.5e-5 rather than the default 1 or 1e-3 from the paper.

**Comment by fchollet** [comment]
What's your complete configuration? I can't reproduce your results by just using VGG19 and reducing the total variation loss.

**Comment by titu1994** [comment]
I used the train script in my repository [Neural Style Transfer](https://github.com/titu1994/Neural-Style-Transfer), in which the Network.py script is a duplicate of the Keras example with several parameters exposed to allow for modification.

With that script, the default parameters are as below : 
--image_size 400 
--content_weight 0.025 
--style_weight 1.0
--total_variation_weight 8.5E-05 
--style_scale 1 
**--content_layer "conv5_2"** 
--init_image "content"
--pool_type "max" 
--preserve_color "False" 
--model "vgg19"

I did not realize that the default content layer is block5_conv2 instead of the block4_conv2 layer in the original script / paper, which improves the quality of the result. Also, the initialization is not noise, but the content image itself, since it improves the quality of the output image (the output is no longer grainy).

Sorry about that. I did not realize that the different default parameters for my script were so different from the original example.

I do not believe that changing to block4_conv2 will cause the change in the commit to have much effect. At the very least, it would strength the style weight slightly. If any meaningful improvement was to be obtained, it would have been glaringly obvious in the output using the 5_2 layer, since this layer is much more susceptible to different style weights in my experience.

**Comment by fchollet** [comment]
These are some modifications that would be good to incorporate in the example script, specifically:
- use of VGG19
- use of `conv5_2`
- use of content image as initial state for combination

**Comment by sjosund** [comment]
Good points. Since the normalization changes it becomes necessary to also change the default value for style weight. Here's starry night/tubingen using the master branch with the default settings.

master_sw1__at_iteration_9

Here's using the changed branch, and changing the style weight to 10.

fix2sw10__at_iteration_9

**Comment by titu1994** [comment]
@fchollet I agree. I thought of submitting a PR for those changes a year ago, but considering that the script was simply an example, I assumed it would benefit by adhering to the original paper.

There are even more improvements in the INetwork.py script in my repository (such as activation shift of the gram matrix, correlation chaining, geometric layer weight adjustment for style inference and using all the layers of VGG 16 or 19), which improve the quality much more than the simpler original script, but I think that is well out of scope for an example script.

**Comment by sjosund** [comment]
@fchollet Regarding the additional suggested modifications (VGG19/conv5_2/content-as-init), do you think that it should just be changed to those or that they should be configurable via command line args?

**Comment by fchollet** [comment]
The style loss should not be 10. Any changes should be tested on a few images first (more than one). What we should do is apply the changes suggested by @titu1994, specifically:

- use of VGG19
- use of conv5_2
- use of content image as initial state for combination

**Comment by titu1994** [comment]
@fchollet I've created a PR https://github.com/fchollet/keras/pull/6877, which incorporates the changes which you have mentioned.

**Comment by fchollet** [comment]
Overall, there doesn't seem to be any change in result quality with the normalization factor change proposed in this PR. We won't be merging. Thanks for the PR and the discussion.

**Comment by sjosund** [comment]
Thanks for the discussion and feedback.
